pkg/util/gitutil/git.go

```diff
@@ -55,48 +90,119 @@ func GetGitRepository(dir string) (*git.Repository, error) {
-	const GitHubHTTPSPrefix = "https://github.com/"
-	const GitHubRepositorySuffix = ".git"
+	return remoteURL, nil
+}
 
-	if strings.HasPrefix(remoteURL, GitHubSSHPrefix) {
-		project = trimGitRemoteURL(remoteURL, GitHubSSHPrefix, GitHubRepositorySuffix)
-	} else if strings.HasPrefix(remoteURL, GitHubHTTPSPrefix) {
-		project = trimGitRemoteURL(remoteURL, GitHubHTTPSPrefix, GitHubRepositorySuffix)
+// IsGitOriginURLGitHub returns true if the provided remoteURL is detected as GitHub.
+func IsGitOriginURLGitHub(remoteURL string) bool {
+	return strings.Contains(remoteURL, GitHubHostName)
+}
+
+// TryGetVCSInfo attempts to detect whether the provided remoteURL
+// is an SSH or an HTTPS remote URL. It then extracts the repo, owner name,
+// and the type (kind) of VCS from it.
+func TryGetVCSInfo(remoteURL string) (*VCSInfo, error) {
+	project := ""
+	vcsKind := ""
+
+	// If the remote is using git SSH, then we extract the named groups by matching
+	// with the pre-compiled regex pattern.
+	if strings.HasPrefix(remoteURL, "git@") {
+		// Most cloud-hosted VCS have the ssh URL of the format git@somehostname.com:owner/repo
+		if cloudSourceControlSSHRegex.MatchString(remoteURL) {
+			groups := getMatchedGroupsFromRegex(cloudSourceControlSSHRegex, remoteURL)
+			vcsKind = groups["host_name"]
+			project = groups["owner_and_repo"]
+			project = strings.TrimSuffix(project, defaultGitCloudRepositorySuffix)
+		} else if azureSourceControlSSHRegex.MatchString(remoteURL) {
+			// Azure's DevOps service uses a git SSH url, that is completely different
+			// from the rest of the services.
+			groups := getMatchedGroupsFromRegex(azureSourceControlSSHRegex, remoteURL)
+			vcsKind = groups["host_name"]
+			project = groups["owner_and_repo"]
+			project = strings.TrimSuffix(project, defaultGitCloudRepositorySuffix)
+		}
+	} else if strings.HasPrefix(remoteURL, "http") {
+		// This could be an HTTP(S)-based remote.
+		if parsedURL, err := url.Parse(remoteURL); err == nil {
+			vcsKind = parsedURL.Host
+			project = parsedURL.Path
+			// Replace the .git extension from the path.
+			project = strings.TrimSuffix(project, defaultGitCloudRepositorySuffix)
+			// Remove the prefix "/". TrimPrefix returns the same value if there is no prefix.
+			// So it is safe to use it instead of doing any sort of substring matches.
+			project = strings.TrimPrefix(project, "/")
+		}
+	}
+
+	if project == "" {
+		return nil, errors.Errorf("detecting the VCS info from the remote URL %v", remoteURL)
 	}
 
+	// At this point, project is either an empty string or contains the information that we want.
+	// It is OK to run the split on the empty string anyway.
 	split := strings.Split(project, "/")
 
-	if len(split) != 2 {
-		return "", "", errors.Errorf("could not detect GitHub project from url: %v", remote)
+	// For Azure, we will have more than 2 parts in the array.
+	// Ex: owner/project/repo.git
+	if len(split) > 2 {
+		azureSplit := strings.SplitN(project, "/", 2)
+		return &VCSInfo{
```

Comment:
It's better to have no data than incorrect/wrong data. So if we are using an Azure-specific heuristic to obtain the owner and repo, we should confirm that `vcsKind` is actually Azure. (Since otherwise we are blindly making assumptions about the project name, and how many slashes it has, etc. Since if `strings.SplintN` returns `[]` this would panic.